### DISCUSSION TITLE
Pre-build step of a project executing before building its dependency?

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [dotnet/msbuild#10802] [Q&A]
I have the following use case in an msbuild solution:
1. One project builds an assembler
2. Another project uses that assembler in its pre-build step to assemble a file
3. Project 2 has project 1 as one of its ProjectReferences to ensure the assembler is built before it's used

The relevant portion of 2's csproj is this:
```
    <ItemGroup>
      <ProjectReference Include="..\FtRandoLib\FtRandoLib.csproj" />
      <ProjectReference Include="..\snarfblasm\snarfblasm.csproj">
        <ReferenceOutputAssembly>False</ReferenceOutputAssembly>
      </ProjectReference>
    </ItemGroup>
    <ItemGroup>
      <None Update="Resources\mm2rng_prepatch.asm">
        <Generator></Generator>
      </None>
    </ItemGroup>
    <Target Name="PreBuild" BeforeTargets="PreBuildEvent">
      <Exec Command="$(SolutionDir)snarfblasm/$(OutputPath)/snarfblasm.exe $(ProjectDir)Resources/mm2rng_prepatch.asm $(ProjectDir)Resources/mm2rng_prepatch.ips -checking:on -newer:on" />
    </Target>
```
This works fine on my own computer in Visual Studio 2022. However, when this is executed as part of a GitHub CI action (while this has so far only been seen in GitHub actions, `dotnet build` appears to all be a single step in the action, leading me to believe this reflects msbuild behavior; action files [here](https://github.com/TheRealQuantam/MegaMan2Randomizer/tree/snarfblasm-prepatch2/.github/workflows)), the build fails for:
```
  /usr/bin/sh: 2: /tmp/MSBuildTemprunner/tmp31ae2027edec4a3d9c72bbf6e74d52c6.exec.cmd: /home/runner/work/MegaMan2Randomizer2/MegaMan2Randomizer2/snarfblasm/bin/Debug/net8.0//snarfblasm.exe: not found
```
This failure occurs *before* the assembler is even built (in fact it's one of the first lines of the [build log](https://github.com/squid-man/MegaMan2Randomizer2/actions/runs/11305009085/job/31444036582?pr=206#logs)). Furthermore, it happens regardless of whether it's running on Windows or Linux.

Does msbuild allow pre-build actions to occur prior to building projects that are dependencies of the project with the pre-build step?

**Comment by JanKrivanek** [comment]
Try to collect binlog for the success and failed case and it should help you pinpoint the difference.
Binlog viewer should help you show the targets sequencing and reason for them. Is the `PreBuildEvent` defined and executed? 
Alternatively post the binlog here - provided it doesn't contain any sensitive information (otherwise you'd need to post privately via visual studio community issues)
